### ISSUE TITLE
ffi-safe lint should be more aggressive in checking ffi stability of sub-types

### ISSUE THREAD

**Issue by fschutt** [rust-lang/rust#72939]
According to [this answer](https://internals.rust-lang.org/t/question-about-c-abi-stability/12449/4?u=sharazam), this code:

```rust
pub struct A { a: usize, b: usize }

#[repr(C)] 
pub struct B { 
    a: A 
}
```

... should emit a warning that `struct A` is still `#[repr(Rust)]`, so the layout of `B` isn't completely defined, specifically the order of the fields `a` and `b` are not specified in `#[repr(Rust)]`. I'm not sure if the `ffi-safe` lint is active or not, but it would be very useful, so I know that if a type is marked `#[repr(C)]`, all sub-types are also marked `#[repr(C)]`.

Another question: When compiling to a `cdylib`, why does cargo use `#[repr(Rust)]` for publicly visible structs of a crate. It would be nice if, when compiling to a `cdylib` with this code:

```rust
pub fn foo(b: B) -> A { b.a }
```

... there could be automatic warnings:

- `foo` is a public function, should be marked as `#[no_mangle]`
- `B` and `A` should be marked as `#[repr(C)]`

... because the only representation that you usually want when compiling to a `cdylib` is `#[repr(C)]` (or `#[repr(C, u8)]` for enums).
